payment_gateway_adyen/services/payment_service.py

```diff
@@ -0,0 +1,174 @@
+    'Refused': 'failed',
+    'RedirectShopper': 'pending',
+    '[capture-received]': 'succeeded'}
+
+# decimal points of currencies https://docs.adyen.com/developers/currency-codes
+ZERO_DECIMAL_CURRENCIES = [
+    u'CVE', u'DJF', u'GNF', u'IDR', u'JPY', u'KMF', u'KRW', u'PYG',
+    u'RWF', u'UGX', u'VND', u'VUV', u'XAF', u'XOF', u'XPF',
+],
+
+THREE_DECIMAL_CURRENCIES = [
+    u'BHD', u'JOD', u'KWD', u'LYD', u'OMR', u'TND'
+]
+
+
+class PaymentService(Component):
+    _inherit = 'payment.service'
+    _name = 'payment.service.adyen'
+    _usage = 'gateway.provider'
+    _allowed_capture_method = ['immediately']
+
+    def process_return(self, **params):
+        payload = {}
+        payload["browserInfo"] = {
+            "userAgent": "Mozilla/5.0 (X11; U; Linux i686; en-US; rv:\
+            1.9) Gecko/2008052912 Firefox/3.0",
+            "acceptHeader": "text/html,application/xhtml+xml,\
+            application/xml;q=0.9,*/*;q=0.8"
+        }
+        payload["md"] = params['MD']
+        payload["paResponse"] = params['PaRes']
+        transaction = None
+        try:
+            result = self._get_adyen_client().authorise3d(request=payload)
+            vals = {
+                'state': MAP_SOURCE_STATE[result.message['resultCode']],
+                'data': json.dumps(result.message),
+                }
+
+            transaction = self.env['gateway.transaction'].search([
+                ('external_id', '=', result.message['pspReference']),
+                ('payment_mode_id.provider', '=', 'adyen'),
+                ])
+        except ValueError as e:
+            # NOTE catch orther errors?
+            # do we want to cancel the transaction? Can we do it?
+            raise UserError(e.message)
+
+        if transaction:
+            transaction.write(vals)
+        else:
+            raise UserError(
+                _('The transaction %s do not exist in Odoo') % result.message[
+                    'pspReference'])
+
+    def _get_formatted_amount(self, amount=None):
+        if amount is None:
+            amount = self.collection._get_amount_to_capture()
+        if self.collection.currency_id.name in ZERO_DECIMAL_CURRENCIES:
+            return int(amount)
+        elif self.collection.currency_id.name in THREE_DECIMAL_CURRENCIES:
+            return int(float_round(amount * 1000, 0))
+        else:
+            return int(float_round(amount * 100, 0))
+
+    def _prepare_charge(self, source=None, **kwargs):
+        transaction = self.collection
+        description = "|".join([
+            transaction.name,
+            transaction.partner_id.email,
+            str(transaction.id)])
+        vals = {
+            'amount': {
+                "value": self._get_formatted_amount(),
+                "currency": transaction.currency_id.name},
+            'reference': description,
+            'additionalData': {"executeThreeD": "true"}  # TODO optional?
+            }
+        if kwargs.get('card'):
```

Comment:
We should remove this case or add a test case.
In anycase no body will use clear PCI as the alternative is simple and work perfectly. So I prefere to remove it

Reply:
untill recently, the test cases were testing this. I can add a test-case again but the issue is that it will be hard to replay it because it will rely on an unlocked PCI test account, which requires to ask the support to unlock it. The only issue I can see with removing this is may be it could make the python 3 port harder if the unofficial adyen-cse-python lib is that if it doesn't work well on Python3 for instance (I don't know).

Reply:
so I tested and with Python 3, the CSE lib can be imported but there is some minor unicode error. I propose to comment the 3 lines of PCI enabled code for documentation to help the porting process eventually.